src/server/bitops_family.cc

```diff
@@ -181,60 +253,237 @@ bool SetBitValue(uint32_t offset, bool bit_value, std::string* entry) {
+      element_access.Commit(existing_entry);
     }
   }
   return old_value;
 }
 
+// ---------------------------------------------------------
+
+std::string RunBitOperationOnValues(std::string_view op, const BitsStrVec& values) {
+  // This function accept an operation (either OR, XOR, NOT or OR), and run bit operation
+  // on all the values we got from the database. Note that in case that one of the values
+  // is shorter than the other it would return a 0 and the operation would continue
+  // until we ran the longest value. The function will return the resulting new value
+  std::size_t max_len = 0;
+  std::size_t max_len_index = 0;
+
+  const auto BitOperation = [&]() {
+    if (op == OR_OP_NAME) {
+      std::string default_str{values[max_len_index]};
+      return BitOpString(OrOp, SkipOr, std::move(values), std::move(default_str));
+    } else if (op == XOR_OP_NAME) {
+      return BitOpString(XorOp, SkipXor, std::move(values), std::string(max_len, 0));
+    } else if (op == AND_OP_NAME) {
+      return BitOpString(AndOp, SkipAnd, std::move(values), std::string(max_len, 0));
+    } else if (op == NOT_OP_NAME) {
+      return BitOpNotString(values[0]);
+    } else {
+      LOG(FATAL) << "Operation not supported '" << op << "'";
+      return std::string{};  // otherwise we will have warning of not returning value
+    }
+  };
+
+  if (values.empty()) {  // this is ok in case we don't have the src keys
+    return std::string{};
+  }
+  // The new result is the max length input
+  max_len = values[0].size();
+  for (std::size_t i = 1; i < values.size(); ++i) {
+    if (values[i].size() > max_len) {
+      max_len = values[i].size();
+      max_len_index = i;
+    }
+  }
+  return BitOperation();
+}
+
+OpResult<std::string> CombineResultOp(ShardStringResults result, std::string_view op) {
+  // take valid result for each shard
+  BitsStrVec values;
+  int index = 0;
+  for (auto&& res : result) {
+    if (res) {
+      auto v = res.value();
+      values.emplace_back(std::move(v));
+    } else {
+      if (res.status() != OpStatus::SKIPPED) {  // something went wrong, just bale out
+        return res;
+      }
+      ++index;
+    }
+  }
+  // and combine them to single result
+  return RunBitOperationOnValues(op, values);
+}
+
+// For bitop not - we cannot accumulate
+OpResult<std::string> RunBitOpNot(const OpArgs& op_args, ArgSlice keys) {
+  DCHECK(keys.size() == 1);
+
+  EngineShard* es = op_args.shard;
+  // if we found the value, just return, if not found then skip, otherwise report an error
+  auto key = keys.front();
+  OpResult<PrimeIterator> find_res = es->db_slice().Find(op_args.db_cntx, key, OBJ_STRING);
+  if (find_res) {
+    return GetString(es, find_res.value()->second);
+  } else {
+    if (find_res.status() == OpStatus::KEY_NOTFOUND) {
+      return OpStatus::SKIPPED;
+    } else {
+      return find_res.status();
+    }
+  }
+}
+// Read only operation where we are running hte bit operation on all the
```

Comment:
empty line before the comment. also `hte` -> `the`

Reply:
added